Errors reported only from the first failing linter

**Describe the bug**

I decided to try using go-based treefmt to see if it doesn't have issues with formatting hidden files and I noticed that while it now works with hidden files, it seems to report failures for only the first formatter that fails.

**To Reproduce**

Steps to reproduce the behaviour:

1. use this contrived `treefmt.toml` (this was generated by `treefmt-nix`, so you'd have to adjust paths for your system):

   ```toml
   [formatter]
   [formatter.actionlint]
   command =  "/nix/store/r9h133c9m8f6jnlsqzwf89zg9w0w78s8-bash-5.2-p15/bin/bash"
   excludes = []
   includes = [".github/workflows/*.yml", "./github/workflows/*.yml"]
   options = ["-eucx", "sleep 2; /nix/store/gksvgsiwzwc69lm7rm54wgfll26c2dyp-actionlint-1.6.26/bin/actionlint"]
   
   [formatter.ruff-check]
   command = "/nix/store/r9h133c9m8f6jnlsqzwf89zg9w0w78s8-bash-5.2-p15/bin/bash"
   excludes = ["apks/chromium/*"]
   includes = ["*.py", "*.pyi"]
   options = ["-eucx", "sleep 1; /nix/store/372fyydc1xp2mjj1wwd8n28qlp0fp5dg-ruff-0.1.6/bin/ruff check ."]
   ```
3. run `treefmt` and observe you see errors from `ruff` only,
2. switch the `sleep` durations around,
4. run `treefmt` and observe you see errors from `actionlint` only.

**Expected behaviour**

Errors from all failing formatters/linters are reported.

**System information**

I used `nixpkgs-unstable#treefmt2` which at this point in time points to `2.0.0-rc4`.

@jaen I tried the steps above, and I don't think I'm seeing what you were seeing:

```console
❯ nix fmt -- --no-cache 
ruff-check error:
flavors/lineageos/update_device_dirs.py:196:82: E703 [*] Statement ends with an unnecessary semicolon
flavors/lineageos/update_device_metadata.py:7:8: F401 [*] `urllib.request` imported but unused
Found 2 errors.
[*] 2 fixable with the `--fix` option.

actionlint error:
.github/workflows/grapheneos-update.yml:16:7: shellcheck reported issue in this script: SC2086:info:5:52: Double quote to prevent globbing and word splitting [shellcheck]
   |
16 |       run: |
   |       ^~~~
.github/workflows/grapheneos-update.yml:16:7: shellcheck reported issue in this script: SC2086:info:10:52: Double quote to prevent globbing and word splitting [shellcheck]
   |
16 |       run: |
   |       ^~~~
.github/workflows/instantiate.yml:17:7: shellcheck reported issue in this script: SC2046:warning:1:98: Quote this to prevent word splitting [shellcheck]
   |
17 |     - run: |
   |       ^~~~
.github/workflows/instantiate.yml:17:7: shellcheck reported issue in this script: SC2071:error:3:32: > is for string comparisons. Use -gt instead [shellcheck]
   |
17 |     - run: |
   |       ^~~~
.github/workflows/instantiate.yml:26:7: shellcheck reported issue in this script: SC2086:info:4:8: Double quote to prevent globbing and word splitting [shellcheck]
   |
26 |     - run: |
   |       ^~~~
.github/workflows/instantiate.yml:56:7: shellcheck reported issue in this script: SC2046:warning:3:4: Quote this to prevent word splitting [shellcheck]
   |
56 |     - run: |
   |       ^~~~
.github/workflows/instantiate.yml:56:7: shellcheck reported issue in this script: SC2086:info:4:19: Double quote to prevent globbing and word splitting [shellcheck]
   |
56 |     - run: |
   |       ^~~~
.github/workflows/instantiate.yml:56:7: shellcheck reported issue in this script: SC2086:info:5:43: Double quote to prevent globbing and word splitting [shellcheck]
   |
56 |     - run: |
   |       ^~~~

robotnix on  HEAD (d8a35e1) [!?] on ☁️   took 1m1s 
❯ nix fmt -- --no-cache
robotnix on  HEAD (d8a35e1) [!?] on ☁️   took 1m1s 
❯ nix fmt -- --no-cache 
WARN format: no formatter for path: flavors/lineageos/update_device_dirs.py
WARN format: no formatter for path: flavors/lineageos/update_device_metadata.py
WARN format: no formatter for path: flavors/waydroid/extract-patch-metadata.py
WARN format: no formatter for path: modules/apps/chromium-trichrome-patcher.py
WARN format: no formatter for path: modules/apv/autogenerate.py
WARN format: no formatter for path: scripts/mk_repo_file.py
WARN format: no formatter for path: scripts/robotnix_common.py
WARN format: no formatter for path: scripts/test_mk_repo_file.py
actionlint error:
.github/workflows/grapheneos-update.yml:16:7: shellcheck reported issue in this script: SC2086:info:5:52: Double quote to prevent globbing and word splitting [shellcheck]
   |
16 |       run: |
   |       ^~~~
.github/workflows/grapheneos-update.yml:16:7: shellcheck reported issue in this script: SC2086:info:10:52: Double quote to prevent globbing and word splitting [shellcheck]
   |
16 |       run: |
   |       ^~~~
.github/workflows/instantiate.yml:17:7: shellcheck reported issue in this script: SC2046:warning:1:98: Quote this to prevent word splitting [shellcheck]
   |
17 |     - run: |
   |       ^~~~
.github/workflows/instantiate.yml:17:7: shellcheck reported issue in this script: SC2071:error:3:32: > is for string comparisons. Use -gt instead [shellcheck]
   |
17 |     - run: |
   |       ^~~~
.github/workflows/instantiate.yml:26:7: shellcheck reported issue in this script: SC2086:info:4:8: Double quote to prevent globbing and word splitting [shellcheck]
   |
26 |     - run: |
   |       ^~~~
.github/workflows/instantiate.yml:56:7: shellcheck reported issue in this script: SC2046:warning:3:4: Quote this to prevent word splitting [shellcheck]
   |
56 |     - run: |
   |       ^~~~
.github/workflows/instantiate.yml:56:7: shellcheck reported issue in this script: SC2086:info:4:19: Double quote to prevent globbing and word splitting [shellcheck]
   |
56 |     - run: |
   |       ^~~~
.github/workflows/instantiate.yml:56:7: shellcheck reported issue in this script: SC2086:info:5:43: Double quote to prevent globbing and word splitting [shellcheck]
   |
56 |     - run: |
   |       ^~~~

treefmt: error: formatting failure: formatter '/nix/store/gksvgsiwzwc69lm7rm54wgfll26c2dyp-actionlint-1.6.26/bin/actionlint' with options '[]' failed to apply: exit status 1
```

I think you were right and this needs a simpler reproducer.

I spent some time improving #366. 

Here's an updated version of the reproducer using it: [treefmt-fail-fix.tar.gz](https://github.com/user-attachments/files/16850537/treefmt-fail-fix.tar.gz)

On the first application, you can see that `ruff-check` fails, but it proceeds with `ruff-format`. The file `src/nix/sum.py` is detected as having changed, however we do not update the cache, to ensure this file is re-processed on subsequent runs in case the problem was something like a config error which is easily fixed.

```console
❯ nix fmt -- -vv 
DEBU format: config-file=/nix/store/0f4l3y7lkfcnqriahmdlh01xvcf7qby7-treefmt.toml tree-root=/home/brian/Downloads/treefmt-fail
DEBU cache: finished generating change set in 106.491µs
DEBU format | nixpkgs-fmt: match: /home/brian/Downloads/treefmt-fail/src/nix/sum.nix
DEBU format | ruff-check: match: /home/brian/Downloads/treefmt-fail/src/nix/sum.py
DEBU format | ruff-format: match: /home/brian/Downloads/treefmt-fail/src/nix/sum.py
DEBU format | ruff-check: executing: /nix/store/s9sn9llf7xk4hjac2d4phj5zha9gszyh-ruff-0.6.3/bin/ruff check src/nix/sum.py
DEBU format | nixpkgs-fmt: executing: /nix/store/vfmlh00mb40qarwj236vmkqylcnv5la5-nixfmt-unstable-2024-08-16/bin/nixfmt src/nix/sum.nix
ERRO format | ruff-check: failed to apply with options '[check]': exit status 1
ruff-check error:
src/nix/sum.py:10:8: F401 [*] `urlib.request` imported but unused
   |
 8 | import os
 9 | import pathlib
10 | import urlib.request
   |        ^^^^^^^^^^^^^ F401
11 | 
12 | from typing import Any, Callable, Dict, List, Optional, cast
   |
   = help: Remove unused import: `urlib.request`

src/nix/sum.py:166:54: E703 [*] Statement ends with an unnecessary semicolon
    |
164 |     device_dirs_fn = os.path.join(args.branch, 'device-dirs.json')
165 |     if os.path.exists(device_dirs_fn):
166 |         device_dirs = json.load(open(device_dirs_fn));
    |                                                      ^ E703
167 |     else:
168 |         device_dirs = {}
    |
    = help: Remove unnecessary semicolon

Found 2 errors.
[*] 2 fixable with the `--fix` option.

DEBU format | ruff-format: executing: /nix/store/s9sn9llf7xk4hjac2d4phj5zha9gszyh-ruff-0.6.3/bin/ruff format src/nix/sum.py
INFO format | ruff-format: 1 file(s) processed in 8.664883ms
DEBU format: file has changed path=src/nix/sum.py prev_size=6249 prev_mod_time="2024-09-03 15:39:26 +0100 IST" current_size=6206 current_mod_time="2024-09-03 15:42:46 +0100 IST"
INFO format | nixpkgs-fmt: 1 file(s) processed in 32.92821ms
DEBU format: file has changed path=src/nix/sum.nix prev_size=19626 prev_mod_time="2024-09-03 15:39:26 +0100 IST" current_size=21639 current_mod_time="2024-09-03 15:42:46 +0100 IST"
DEBU cache: finished processing 1 paths in 12.636895ms
traversed 4 files
emitted 2 files for processing
formatted 2 files (2 changed) in 46ms
```

Running it again yields:

```console
❯ nix fmt -- -vv 
DEBU format: config-file=/nix/store/0f4l3y7lkfcnqriahmdlh01xvcf7qby7-treefmt.toml tree-root=/home/brian/Downloads/treefmt-fail
DEBU cache: finished generating change set in 137.501µs
DEBU format | ruff-check: match: /home/brian/Downloads/treefmt-fail/src/nix/sum.py
DEBU format | ruff-format: match: /home/brian/Downloads/treefmt-fail/src/nix/sum.py
DEBU format | ruff-check: executing: /nix/store/s9sn9llf7xk4hjac2d4phj5zha9gszyh-ruff-0.6.3/bin/ruff check src/nix/sum.py
ERRO format | ruff-check: failed to apply with options '[check]': exit status 1
ruff-check error:
src/nix/sum.py:10:8: F401 [*] `urlib.request` imported but unused
   |
 8 | import os
 9 | import pathlib
10 | import urlib.request
   |        ^^^^^^^^^^^^^ F401
11 | 
12 | from typing import Any, Callable, Dict, List, Optional, cast
   |
   = help: Remove unused import: `urlib.request`

Found 1 error.
[*] 1 fixable with the `--fix` option.

DEBU format | ruff-format: executing: /nix/store/s9sn9llf7xk4hjac2d4phj5zha9gszyh-ruff-0.6.3/bin/ruff format src/nix/sum.py
INFO format | ruff-format: 1 file(s) processed in 7.4775ms
DEBU cache: finished processing 0 paths in 610ns
traversed 4 files
emitted 1 files for processing
formatted 1 files (0 changed) in 18ms
```

`ruff-check` still fails, and `ruff-format` is still applied, but this time there are no changes detected as there was nothing for `ruff-format` to do. 

If you then add this to `flake.nix` and run again:

```nix
settings.formatter.ruff-check.options = nixpkgs.lib.mkForce [
    "check"
    "--fix"
];
```

You see that the python file is re-processed, and check now passes:

```console
❯ nix fmt -- -vv 
DEBU format: config-file=/nix/store/masnfl7cfhpv7jqk0qmx14zvzp65a4kl-treefmt.toml tree-root=/home/brian/Downloads/treefmt-fail
DEBU cache: finished generating change set in 148.382µs
DEBU format | nixpkgs-fmt: match: /home/brian/Downloads/treefmt-fail/flake.nix
DEBU format | ruff-format: match: /home/brian/Downloads/treefmt-fail/src/nix/sum.py
DEBU format | ruff-check: match: /home/brian/Downloads/treefmt-fail/src/nix/sum.py
DEBU format | ruff-check: executing: /nix/store/s9sn9llf7xk4hjac2d4phj5zha9gszyh-ruff-0.6.3/bin/ruff check --fix src/nix/sum.py
DEBU format | nixpkgs-fmt: executing: /nix/store/vfmlh00mb40qarwj236vmkqylcnv5la5-nixfmt-unstable-2024-08-16/bin/nixfmt flake.nix
INFO format | ruff-check: 1 file(s) processed in 10.150282ms
DEBU format | ruff-format: executing: /nix/store/s9sn9llf7xk4hjac2d4phj5zha9gszyh-ruff-0.6.3/bin/ruff format src/nix/sum.py
INFO format | nixpkgs-fmt: 1 file(s) processed in 12.435466ms
INFO format | ruff-format: 1 file(s) processed in 8.326802ms
DEBU format: file has changed path=src/nix/sum.py prev_size=6206 prev_mod_time="2024-09-03 15:50:37 +0100 IST" current_size=6185 current_mod_time="2024-09-03 15:50:49 +0100 IST"
DEBU cache: finished processing 2 paths in 12.401286ms
traversed 4 files
emitted 2 files for processing
formatted 2 files (1 changed) in 32ms
```

I still need to add a test before this PR is ready for merging. I'll try and get it wrapped up in the next couple days.

--- a/cli/cli.go
+++ b/cli/cli.go
@@ -43,8 +43,8 @@ type Format struct {
 	globalExcludes []glob.Glob
 
 	filesCh     chan *walk.File
-	formattedCh chan *walk.File
-	processedCh chan *walk.File
+	formattedCh chan *format.Task
+	processedCh chan *format.Task
 }
 
 func (f *Format) configureLogging() {

--- a/cli/format.go
+++ b/cli/format.go
@@ -184,10 +184,10 @@ func (f *Format) Run() (err error) {
 	f.filesCh = make(chan *walk.File, BatchSize*runtime.NumCPU())
 
 	// create a channel for files that have been formatted
-	f.formattedCh = make(chan *walk.File, cap(f.filesCh))
+	f.formattedCh = make(chan *format.Task, cap(f.filesCh))
 
 	// create a channel for files that have been processed
-	f.processedCh = make(chan *walk.File, cap(f.filesCh))
+	f.processedCh = make(chan *format.Task, cap(f.filesCh))
 
 	// start concurrent processing tasks in reverse order
 	eg.Go(f.updateCache(ctx))
@@ -317,17 +317,21 @@ func (f *Format) applyFormatters(ctx context.Context) func() error {
 
 			// asynchronously apply the sequence formatters to the batch
 			fg.Go(func() error {
-				// iterate the formatters, applying them in sequence to the batch of tasks
-				// we get the formatters list from the first task since they have all the same formatters list
-				for _, f := range tasks[0].Formatters {
-					if err := f.Apply(ctx, tasks); err != nil {
-						return err
+				// Iterate the formatters, applying them in sequence to the batch of tasks.
+				// We get the formatter list from the first task since they have all the same formatters list.
+				formatters := tasks[0].Formatters
+
+				var formatErrors []error
+				for idx := range formatters {
+					if err := formatters[idx].Apply(ctx, tasks); err != nil {
+						formatErrors = append(formatErrors, err)
 					}
 				}
 
 				// pass each file to the formatted channel
 				for _, task := range tasks {
-					f.formattedCh <- task.File
+					task.Errors = formatErrors
+					f.formattedCh <- task
 				}
 
 				return nil
@@ -359,7 +363,9 @@ func (f *Format) applyFormatters(ctx context.Context) func() error {
 			if format.PathMatches(file.RelPath, f.globalExcludes) {
 				log.Debugf("path matched global excludes: %s", file.RelPath)
 				// mark it as processed and continue to the next
-				f.formattedCh <- file
+				f.formattedCh <- &format.Task{
+					File: file,
+				}
 				continue
 			}
 
@@ -378,7 +384,9 @@ func (f *Format) applyFormatters(ctx context.Context) func() error {
 				}
 				log.Logf(f.OnUnmatched, "no formatter for path: %s", file.RelPath)
 				// mark it as processed and continue to the next
-				f.formattedCh <- file
+				f.formattedCh <- &format.Task{
+					File: file,
+				}
 			} else {
 				// record the match
 				stats.Add(stats.Matched, 1)
@@ -414,14 +422,15 @@ func (f *Format) detectFormatted(ctx context.Context) func() error {
 			// detect ctx cancellation
 			case <-ctx.Done():
 				return ctx.Err()
-			// take the next file that has been processed
-			case file, ok := <-f.formattedCh:
+			// take the next task that has been processed
+			case task, ok := <-f.formattedCh:
 				if !ok {
 					// channel has been closed, no further files to process
 					return nil
 				}
 
 				// check if the file has changed
+				file := task.File
 				changed, newInfo, err := file.HasChanged()
 				if err != nil {
 					return err
@@ -451,7 +460,7 @@ func (f *Format) detectFormatted(ctx context.Context) func() error {
 				}
 
 				// mark as processed
-				f.processedCh <- file
+				f.processedCh <- task
 			}
 		}
 	}
@@ -460,12 +469,16 @@ func (f *Format) detectFormatted(ctx context.Context) func() error {
 func (f *Format) updateCache(ctx context.Context) func() error {
 	return func() error {
 		// used to batch updates for more efficient txs
-		batch := make([]*walk.File, 0, BatchSize)
+		batch := make([]*format.Task, 0, BatchSize)
 
 		// apply a batch
 		processBatch := func() error {
 			// pass the batch to the cache for updating
-			if err := cache.Update(batch); err != nil {
+			files := make([]*walk.File, len(batch))
+			for idx := range batch {
+				files[idx] = batch[idx].File
+			}
+			if err := cache.Update(files); err != nil {
 				return err
 			}
 			batch = batch[:0]
@@ -486,11 +499,13 @@ func (f *Format) updateCache(ctx context.Context) func() error {
 			case <-ctx.Done():
 				return ctx.Err()
 			// respond to formatted files
-			case file, ok := <-f.processedCh:
+			case task, ok := <-f.processedCh:
 				if !ok {
 					// channel has been closed, no further files to process
 					break LOOP
 				}
+
+				file := task.File
 
 				if f.Stdin {
 					// dump file into stdout
@@ -508,11 +523,16 @@ func (f *Format) updateCache(ctx context.Context) func() error {
 					continue
 				}
 
-				// append to batch and process if we have enough
-				batch = append(batch, file)
-				if len(batch) == BatchSize {
-					if err := processBatch(); err != nil {
-						return err
+				// Append to batch and process if we have enough.
+				// We do not cache any files that were part of a pipeline in which one or more formatters failed.
+				// This is to ensure those files are re-processed in later invocations after the user has potentially
+				// resolved the issue, e.g. fixed a config problem.
+				if len(task.Errors) == 0 {
+					batch = append(batch, task)
+					if len(batch) == BatchSize {
+						if err := processBatch(); err != nil {
+							return err
+						}
 					}
 				}
 			}

--- a/docs/contributing.md
+++ b/docs/contributing.md
@@ -37,8 +37,6 @@ updated `nix/packages/treefmt/gomod2nix.toml`.
 
 To sync it up, run `nix develop .#renovate -c gomod2nix:update`.
 
-
-
 ## Making changes
 
 If you want to introduce changes to the project, please follow these steps:

--- a/format/formatter.go
+++ b/format/formatter.go
@@ -75,15 +75,14 @@ func (f *Formatter) Apply(ctx context.Context, tasks []*Task) error {
 	f.log.Debugf("executing: %s", cmd.String())
 
 	if out, err := cmd.CombinedOutput(); err != nil {
+		f.log.Errorf("failed to apply with options '%v': %s", f.config.Options, err)
 		if len(out) > 0 {
 			_, _ = fmt.Fprintf(os.Stderr, "%s error:\n%s\n", f.name, out)
 		}
 		return fmt.Errorf("formatter '%s' with options '%v' failed to apply: %w", f.config.Command, f.config.Options, err)
+	} else {
+		f.log.Infof("%v file(s) processed in %v", len(tasks), time.Since(start))
 	}
-
-	//
-
-	f.log.Infof("%v file(s) processed in %v", len(tasks), time.Since(start))
 
 	return nil
 }

--- a/format/task.go
+++ b/format/task.go
@@ -11,6 +11,7 @@ type Task struct {
 	File       *walk.File
 	Formatters []*Formatter
 	BatchKey   string
+	Errors     []error
 }
 
 func NewTask(file *walk.File, formatters []*Formatter) Task {

